mlir/test/lib/Dialect/EmitC/TestTypeConversions.cpp

```diff
@@ -0,0 +1,102 @@
+//===- TestTypeConversions.cpp - Test EmitC type conversions --------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file contains a test pass to check EmitC type conversions.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/EmitC/IR/EmitC.h"
+#include "mlir/Dialect/EmitC/Transforms/TypeConversions.h"
+#include "mlir/IR/BuiltinAttributeInterfaces.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/Value.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+using namespace mlir;
+
+namespace {
+struct TestEmitCTypeConversions
+    : public PassWrapper<TestEmitCTypeConversions, OperationPass<>> {
+  MLIR_DEFINE_EXPLICIT_INTERNAL_INLINE_TYPE_ID(TestEmitCTypeConversions)
+
+  void runOnOperation() override;
+  StringRef getArgument() const final { return "test-emitc-type-conversions"; }
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<emitc::EmitCDialect>();
+  }
+  StringRef getDescription() const final {
+    return "Test EmitC type conversions";
+  }
+};
+
+struct ConstantConverter : public OpConversionPattern<emitc::ConstantOp> {
+public:
+  using OpConversionPattern<emitc::ConstantOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(emitc::ConstantOp constantOp, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    // Get value attribute
+    auto valueAttr = adaptor.getValueAttr();
+    auto typedValueAttr = ::cast<TypedAttr>(valueAttr);
+
+    Type dstType = getTypeConverter()->convertType(typedValueAttr.getType());
+
+    if (!dstType)
+      return rewriter.notifyMatchFailure(constantOp, "type conversion failed");
+
+    Attribute convertedAttr = valueAttr;
+
+    if (auto iType = dyn_cast<IndexType>(typedValueAttr.getType())) {
+      auto intValAttr = dyn_cast<IntegerAttr>(valueAttr);
+      auto val = intValAttr.getValue();
+      convertedAttr = IntegerAttr::get(dstType, val);
```

Comment:
Run will fail here because `IntegerAttr` expects IntegerType or IndexType, not our EmitC types. We can figure this out later, we'll need to convert these constants anyway should they show up (`arith.constant`....) so this is not time lost.

Reply:
I think it's fine to have `emitc.constant` that have values of `IntegerAttr` with type `index`, as long as the result type of the `emitc.constant` is `emitc.size_t`. We have the same for array constants, where the values have tensor type.

Reply:
The op itself is checking for a match of the two and complaining if they don't: 
```
'emitc.constant' op requires attribute to either be an #emitc.opaque attribute or it's type ('index') to match the op's result type ('!emitc.usize_t')
```
... so, I'd favor relaxing this rule to just assert the types are "isomorphic", if that makes sense.